### PR TITLE
XLSX Pivot Table Downloads have 'Native' Pivot Table

### LINKED ISSUE #2473: The CSV/XLSX download for a pivot table should be pivoted
-  I am using the Chrome browser.
-  My computer's OS is Mac OSX.
-  My Metabase version is v0.17.0

Currently if you run a query which displays a pivoted table of data in the browser then go to download the file you get a different set of results.  This is confusing and inconsistent for users.  If a table is going to be pivoted it should be that way in the CSV as well.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by adam-james-v**
closes #2473

The xlsx downloads work fine for regular tables. However, on Pivot Table downloads, the resulting file isn't pivoted and is formatted somewhat strangely (it contains additional rows related to totals, has an extra 'pivot-grouping' column).

Now, the downloaded xlsx file contains 2 sheets:
- the pivot sheet containing a 'native' Pivot table whose reference points to the data on the data sheet, and whose columns, rows, and values align with the cols, rows, and measures set up in Metabase
- the data sheet containing the question's rows unpivoted. If you were constructing a pivot table yourself, you would probably start with this shape of data.

I have Numbers locally, and things work correctly there:
image

Uploading the same file to Google Sheets also produces a working Pivot Table:
image

And a video demonstrating interactivity with the native pivot:

https://github.com/metabase/metabase/assets/21064735/36cfffea-5808-4182-8701-06143b2d9046

### CODE REVIEW

File: src/metabase/query_processor/streaming/xlsx.clj [hunk 1]
```diff
@@ -447,38 +478,143 @@
...
+                                                   idx)) indices))
+        ordered-cols                    (vec (m/remove-nth pivot-grouping-key ordered-cols))
+        {:keys [pivot-rows
+                pivot-cols
+                pivot-measures
+                aggregation-functions]} (-> pivot-spec
+                                            (update :pivot-rows idx-shift)
+                                            (update :pivot-cols idx-shift)
+                                            (update :pivot-measures idx-shift)
+                                            (update :aggregation-functions #(vec (m/remove-nth pivot-grouping-key %))))
```
Comment (escherize) [inline src/metabase/query_processor/streaming/xlsx.clj:523]: the destructuring feels a bit unneccesary here, what do you think of doing

```suggestion
       pivot-rows (idx-shift (:pivot-rows pivot-spec))
       ...
```

instead?

File: src/metabase/query_processor/streaming/xlsx.clj [hunk 2]
```diff
@@ -447,38 +478,143 @@
...
+
+;; Possible Functions: https://poi.apache.org/apidocs/dev/org/apache/poi/ss/usermodel/DataConsolidateFunction.html
+;; I'm only including the keys that seem to work for our Pivot Tables as of 2024-06-06
+(def ^:private agg-fn-k->pivot-agg-fn-enum
+  {:sum    DataConsolidateFunction/SUM
+   :avg    DataConsolidateFunction/AVERAGE
+   :min    DataConsolidateFunction/MIN
+   :max    DataConsolidateFunction/MAX
+   :count  DataConsolidateFunction/COUNT
+   :stddev DataConsolidateFunction/STD_DEV})
```
Comment (escherize) [inline src/metabase/query_processor/streaming/xlsx.clj:504]: these keys look familiar, should this be defined together with `col->aggregation-function-key`? Do they need to be kept in sync?

File: src/metabase/query_processor/streaming/xlsx.clj [hunk 3]
```diff
@@ -447,38 +478,143 @@
...
+        ^XSSFPivotTable pivot-table     (.createPivotTable ^XSSFSheet pivot-sheet
+                                                           ^AreaReference area-ref
+                                                           (CellReference. "A1")
+                                                           ^XSSFSheet data-sheet)]
+    (doseq [idx pivot-rows]
+      (.addRowLabel pivot-table idx))
+    (doseq [idx pivot-cols]
+      (.addColLabel pivot-table idx))
+    (doseq [idx pivot-measures]
+      (.addColumnLabel pivot-table (agg-fn-k->pivot-agg-fn-enum (get aggregation-functions idx DataConsolidateFunction/COUNT)) idx))
```
[the remark this answers is not in the document]
Comment (escherize) [reply src/metabase/query_processor/streaming/xlsx.clj:544]: I noticed there weren't any added unit tests, and I think that would be a lot easier than recreating a universe when we need to test certain things

**Review by escherize — CHANGES_REQUESTED** [review]
Nice stuff. I had a few questions about this approach, don't take it as you need to take all of these suggestions, but I'd like to at least hear what you think about those questions.

File: src/metabase/query_processor/streaming/xlsx.clj (hunk 3 above)
[the remark this answers is not in the document]
Comment (adam-james-v) [reply src/metabase/query_processor/streaming/xlsx.clj:544]: While I agree with this idea, the native-pivot function already takes in some data structures, the rows, the pivot-specifications, and some column data. 

From those pieces, it creates the workbook object, sheets, and creates the cell formatting objects too. It's a very imperative-java heavy section of code already, and with the existing xlsx functions in the rest of the namespace (the functions responsible for setting up cell styling and adding rows), I think this function is already taking in a reasonble 'data only' input, and dealing with things in an acceptable way, at least before this namespace is improved more hollistically.

File: src/metabase/query_processor/streaming/xlsx.clj (hunk 2 above)
Comment (adam-james-v) [reply src/metabase/query_processor/streaming/xlsx.clj]: I've consolidated this and the other fn into just one that handles this inside the xlsx namespace.

**Review by escherize — APPROVED** [review]
Looks good!

I might try to combine those 3 atoms into one, but if you don't think it's a good idea, don't feel pressured.

**Comment by iethree** [comment]
@adam-james-v "Addresses" is lovely english, but please use github [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) 🙏

**Pull request merged**